quinn/src/lib.rs

```diff
@@ -712,28 +709,22 @@ impl Connection {
     /// # Panics
     /// - If called more than once on handles to the same connection
     // FIXME: Infallible
-    pub fn close(&self, error_code: u16, reason: &[u8]) -> impl Future<Item = (), Error = ()> {
-        let (send, recv) = oneshot::channel();
-        {
-            let endpoint = &mut *self.0.endpoint.borrow_mut();
+    pub fn close(&self, error_code: u16, reason: &[u8]) {
+        let endpoint = &mut *self.0.endpoint.borrow_mut();
 
-            let pending = endpoint.pending.get_mut(&self.0.handle).unwrap();
-            assert!(
-                pending.closing.is_none(),
-                "a connection can only be closed once"
-            );
-            pending.closing = Some(send);
+        let pending = endpoint.pending.get_mut(&self.0.handle).unwrap();
+        assert!(!pending.closing, "a connection can only be closed once");
```

Comment:
Why is this assert necessary? Can't we just early return out if `closing` is `true` already? Might make for a friendlier API.

Reply:
I tend waffle about friendliness vs. making unintended behavior obvious to its author, but I think you've sold me okay on "transport protocol impls shouldn't take out your process."

Reply:
I do also want to make unintended behavior obvious to the author (we could return a `Result` here), but if there is actually no ill effect then it we may not have to signal it.